walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,71 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
+// Class containing the query method to find available time ranges for a given meeting request
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    // Find optional time ranges for the meeting
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        
+        Collection<TimeRange> mandatoryAvailable = new ArrayList<>();
+        ArrayList<TimeRange> mandatoryBusy = new ArrayList<>();
+        
+        // Iterate over the events and add to the mandatoryBusy array the time ranges of events in which mandatory attendees of the request participate
+        for (Event event : events) {
+            Boolean areAttendeesSetsDisjoint = Collections.disjoint(request.getAttendees(), event.getAttendees());
+            // If the request's attendees set and the event's attendees set aren't disjoint - the event's time range is a busy range
+            if (!areAttendeesSetsDisjoint) {
+                mandatoryBusy.add(event.getWhen());
+            }
+        }
+        Collections.sort(mandatoryBusy,TimeRange.ORDER_BY_START);
+        // If there's no busy range - the whole day is available
+        if(mandatoryBusy.isEmpty()) {
+            TimeRange wholeDay = TimeRange.WHOLE_DAY;
+            addIfRangeIsLongEnough(mandatoryAvailable, wholeDay, request);
+            return mandatoryAvailable;
+        }
+        // If the first busy range is not at the beginning of the day - the first time range is available
+        if (mandatoryBusy.get(0).start() != TimeRange.START_OF_DAY) {
+            TimeRange firstRange = TimeRange.fromStartDuration(TimeRange.START_OF_DAY, mandatoryBusy.get(0).start() - TimeRange.START_OF_DAY);
```

Comment:
Why not TimeRange.fromStartEnd(TimeRange.START_OF_DAY, mandatoryBusy.get(0).end())